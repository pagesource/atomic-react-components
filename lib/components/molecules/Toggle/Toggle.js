/* eslint-disable react/prop-types */
// @flow
/**
 *
 * Toggle
 *
 */
import React, { useState } from 'react';
import styled from 'styled-components';
import { InputVanilla as Input } from '../../atoms/Input';
import Label from '../../atoms/Label';

import styles from './Toggle.style';

const Toggle = ({
  id,
  name,
  type,
  disabled,
  className,
  children,
  labelProps,
  labelOn,
  labelOff,
  labelName,
  ...otherProps
}) => {
  const [toggleOn, handleToggle] = useState(false);

  let toggleLabel = '';
  if (toggleOn) {
    toggleLabel = labelOn;
  } else toggleLabel = labelOff;

  return (
    <div className={className}>
      <span className="labelName">{labelName}</span>
      <section>
        <Label htmlFor={id} {...labelProps}>
          <Input
            id={id}
            name={name}
            type={type}
            disabled={disabled}
            onClick={() => {
              handleToggle(!toggleOn);
            }}
            {...otherProps}
          />
          <span className="slider round" />
        </Label>
        <span className="toggleLabel">{toggleLabel}</span>
      </section>
      {children}
    </div>
  );
};

Toggle.defaultProps = {
  id: 'toggletest1',
  name: 'toggleCheckbox',
  type: 'checkbox',
  disabled: false,
  className: 'toggleSwitch',
  children: '',
  labelName: 'Toggle Label',
  labelOn: 'On',
  labelOff: 'Off',
};

export default styled(Toggle)`
  ${styles};
`;

export { Toggle as ToggleVanilla };
