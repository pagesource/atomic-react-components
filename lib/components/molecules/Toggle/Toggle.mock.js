const defaultCheckbox = {
  type: 'checkbox',
  name: 'test_1',
  id: 'checkbox_test_1',
  value: 'off',
  labelName: 'Default Checkbox',
  labelOn: 'On',
  labelOff: 'Off',
  disabled: false,
};

const DisableCheckbox = {
  type: 'checkbox',
  name: 'test_2',
  id: 'checkbox_test_2',
  value: 'off',
  labelName: 'Disable Checkbox',
  labelOn: 'On',
  labelOff: 'Off',
  disabled: true,
};

export { defaultCheckbox, DisableCheckbox };
