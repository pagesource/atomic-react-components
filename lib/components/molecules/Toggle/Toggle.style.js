import { css } from 'styled-components';

const styles = css`

  ${props => `
       ${props.theme.typography.FONT_DEFAULT};
`};
section{
  ${props => `
    padding-top: ${props.theme.spacing.unit1}
    `};
}
  label {
    position: relative;
    display: inline-block;
    ${props => `
      width: ${props.theme.spacing.unit8};
      height:${props.theme.spacing.unit4};
  `};
    cursor: ${props => (props.disabled ? 'default' : 'pointer')};
  }

  label input {
    opacity: 0;
    width: 0;
    height: 0;
  }
  .slider {
    position: absolute;
    cursor: ${props => (props.disabled ? 'default' : 'pointer')};
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    background-color: var(--grey);
    -webkit-transition: 0.4s;
    transition: 0.4s;
  }

  .slider:before {
    position: absolute;
    content: '';
    ${props => `
      width: ${props.theme.spacing.unit3};
      height:${props.theme.spacing.unit3};
      left: ${props.theme.spacing.unit1};
    `};
    bottom: 4px;
    background-color: var(--white);
    -webkit-transition: 0.4s;
    transition: 0.4s;
  }

  input:checked + .slider {
    background-color: var(--success);
  }

  input:checked + .slider:before {
    ${props => `
    webkit-transform: translateX(${props.theme.spacing.unit3});
    -ms-transform: translateX(${props.theme.spacing.unit3});
    transform: translateX(${props.theme.spacing.unit3});
    `};
    -
  }

  .slider.round {
    border-radius: var(--spacing-7);
  }

  .slider.round:before {
    border-radius: 50%;
  }

  .toggleLabel{
    vertical-align : middle;
    ${props => `
    padding-left: ${props.theme.spacing.unit1}
    `};

  }
`;

export default styles;
