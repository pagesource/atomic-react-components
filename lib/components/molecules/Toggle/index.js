// @flow
export { default } from './Toggle';
export { ToggleVanilla } from './Toggle';
