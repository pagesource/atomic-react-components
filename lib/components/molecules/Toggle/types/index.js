// @flow
import type { Node } from 'react';

export type Props = {
  children: Node,
  id: String,
  name: String,
  type: String,
  disabled: Boolean,
  className?: String,
  labelName: String,
  toggleLabel: String,
  labelOn: String,
  labelOff: String,
};
