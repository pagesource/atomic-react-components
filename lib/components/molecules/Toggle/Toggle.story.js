import React from 'react';
import { storiesOf } from '@storybook/react';
// import { action } from '@storybook/addon-actions';
import { defaultCheckbox } from './Toggle.mock';

// Import Styled Component to showcase variations
import Toggle, { ToggleVanilla } from '.';

// TODO: indentation is not working fine in editor. Need to check for .hbs files
// $FlowFixMe eslint-disable-next-line extra-rules/no-commented-out-code
storiesOf('Molecules', module)
  .addParameters({ jest: ['Toggle', 'ToggleVanilla'] })
  .add('Toggle knobs', () => <ToggleVanilla {...defaultCheckbox} />)
  .add('Toggle', () => <Toggle {...defaultCheckbox} />);
