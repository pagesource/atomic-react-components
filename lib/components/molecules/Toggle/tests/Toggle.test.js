import { shallow } from 'enzyme';

import Toggle from '../index';
// import Input from '../../../atoms/Input/Input';
import defaultCheckbox from '../Toggle.mock';

describe('<Toggle />', () => {
  let ToggleComponent = '';
  beforeEach(() => {
    ToggleComponent = shallow(<Toggle {...defaultCheckbox}>Test</Toggle>);
  });

  test('should be propertags', () => {
    expect(ToggleComponent.find('label')).toBeTruthy();
  });

  test('should be propertags', () => {
    expect(ToggleComponent.find('input')).toBeTruthy();
  });

  test('should have input with label', () => {
    ToggleComponent.setProps({ labelName: 'value' });
    expect(ToggleComponent.find('.labelName').text()).toEqual('Value');
  });

  test('should render correctly', () => {
    expect(ToggleComponent).toMatchSnapshot();
  });

  test('should render a div', () => {
    expect(ToggleComponent.find('div')).toBeTruthy();
  });
});
