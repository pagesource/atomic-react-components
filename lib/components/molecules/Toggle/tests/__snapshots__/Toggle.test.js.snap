// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Toggle /> should render correctly 1`] = `
<StyledComponent
  forwardedComponent={
    Object {
      "$$typeof": Symbol(react.forward_ref),
      "attrs": Array [],
      "componentStyle": ComponentStyle {
        "componentId": "Toggle-ezbk6e-0",
        "isStatic": false,
        "rules": Array [
          [Function],
          ";section{",
          [Function],
          ";}label{position:relative;display:inline-block;",
          [Function],
          ";cursor:",
          [Function],
          ";}label input{opacity:0;width:0;height:0;}.slider{position:absolute;cursor:",
          [Function],
          ";top:0;left:0;right:0;bottom:0;background-color:var(--grey);-webkit-transition:0.4s;transition:0.4s;}.slider:before{position:absolute;content:'';",
          [Function],
          ";bottom:4px;background-color:var(--white);-webkit-transition:0.4s;transition:0.4s;}input:checked + .slider{background-color:var(--success);}input:checked + .slider:before{",
          [Function],
          ";-}.slider.round{border-radius:var(--spacing-7);}.slider.round:before{border-radius:50%;}.toggleLabel{vertical-align:middle;",
          [Function],
          ";}",
          ";",
        ],
      },
      "displayName": "Toggle",
      "foldedComponentIds": Array [],
      "render": [Function],
      "styledComponentId": "Toggle-ezbk6e-0",
      "target": [Function],
      "toString": [Function],
      "usesTheme": true,
      "warnTooManyClasses": [Function],
      "withComponent": [Function],
    }
  }
  forwardedRef={null}
>
  Test
</StyledComponent>
`;
